### DISCUSSION TITLE
Can nginx interface with SocketIO using a unix socket

### DISCUSSION THREAD

**Discussion by eric-gilbertson** [miguelgrinberg/Flask-SocketIO#1602] [Q&A]
I am trying to configure Nginx to communicate with Flask-SocketIO via a Unix socket as outlined in the link below but the Nginx WS requests to uwsgi always time out. I would like to use this approach, if possible, because of the it seems that it would simplify deployment of multiple uwsgi's since it would not require allocation of instance specific ports as outlined  in the 'Using nginx as a WebSocket Reverse Proxy' of the Flask-SocketIO documentation. I would appreciate it if you could comment as to whether Abrahmasen's should work and whether it would be a good approach for deploying multiple uwsgi instances under nginx? 

Thank you very much for your great library and the comprehensive documentation.

https://michaelabrahamsen.com/posts/configuring-uwsgi-and-nginx-for-use-with-flask-socketio

**Comment by miguelgrinberg** [comment]
I would think a unix socket should work exactly as a TCP/IP socket.

**Comment by eric-gilbertson** [comment]
Yes thanks, I agree but I don't see how a SocketIO instance can be instantiated using a unix socket. I'm invoking with:

socketio.run(app, debug=True, use_reloader=False, host='unix:///tmp/kzsuwww.sock')

and because eventlet expects a traditional host & port using a socket value for host results in the following exception:

Traceback (most recent call last):
  File "socketio_app.py", line 11, in <module>
    socketio.run(app, debug=True, use_reloader=False, host='unix:///tmp/kzsuwww.sock')
  File "/home/ericg/src/kzsu-web/venv38/lib/python3.8/site-packages/flask_socketio/__init__.py", line 626, in run
    run_server()
  File "/home/ericg/src/kzsu-web/venv38/lib/python3.8/site-packages/flask_socketio/__init__.py", line 600, in run_server
    addresses = eventlet.green.socket.getaddrinfo(host, port)
  File "/home/ericg/src/kzsu-web/venv38/lib/python3.8/site-packages/eventlet/support/greendns.py", line 539, in getaddrinfo
    raise socket.gaierror(socket.EAI_NONAME, 'No address found')
socket.gaierror: [Errno -2] No address found

Any pointers you could provide as to how I can invoke this with a unix socket would be much appreciated.

**Comment by eric-gilbertson** [comment]
Closing, the Abrahamsen example uses greenlet which does support Unix sockets. I was using eventlet which as far as I can tell cannot be configured to use them.

**Comment by miguelgrinberg** [comment]
I think you are confusing the implementation of the `socketio.run()` method, which is a helper method that is optional, with the support in eventlet for unix sockets, which does exist.

Your `app` Flask instance is a standard WSGI server, you can start it via socket.run() if that works for you, or else you can use any other method. Eventlet's WSGI server can be started on its own ([docs](https://eventlet.net/doc/modules/wsgi.html)), and you can also use Gunicorn with the eventlet worker if you prefer something a bit more high-level.

**Comment by eric-gilbertson** [comment]
What do you think of modifying socketio.run() to accept a Unix socket? This would allow users to use the documented startup flow with a Unix socket. I can submit a PR for this if you are interested in such a change.

**Comment by miguelgrinberg** [comment]
I'm okay with adding that. But you'll need to implement it for all the servers that `socketio.run()` can use, not just eventlet. You can use the [Werkzeug](https://werkzeug.palletsprojects.com/en/1.0.x/serving/#werkzeug.serving.run_simple) implementation as a guide.

**Comment by eric-gilbertson** [comment]
Can you clarify uwsgi based SocketIO case outlined in the  **uWSGI Web Server** section of the documentation? When using this approach should I use the traditional app.run() invocation or would I still wrap it with the SocketIO.run() variant? I ask because with the former I get the "You need to use the eventlet server..." startup exception and in the latter I get an "Address already in use" error. 

Our current deployment uses nginx with multiple uwsgi processes and I'd like to preserve this architecture if possible. Any advice or examples you could provide for doing this would be much appreciated.

**Comment by miguelgrinberg** [comment]
@eric-gilbertson I'm confused about your question. When you use uWSGI you run the uWSGI web server directly, so neither `app.run()` nor `socketio.run()` are used.

**Comment by eric-gilbertson** [comment]
Sorry, never mind that my 'run() comments, they are indeed not run when using usgi. The issue I don't get is how to invoke the eventlet server from within my app's startup code so that I don't get the 'You need to use the eventlet server. ' startup exception, e.g. does it matter what port it listens at? Should I model the invocation after the eventlet startup code in SocketIO's __init__.py? Also, if my socket IO load is low would it be okay to instantiate only a single instance of it when starting the master uwgi process or should each child also have one? Again, any examples or advice you can provide is much appreciated. - Eric

**Comment by miguelgrinberg** [comment]
uWSGI does not work with eventlet. What are you setting `async_mode` to?
